### PR TITLE
dirtree.elections v1.0.0 (R6 object interface, and improved documentation + tests)

### PR DESCRIPTION
**Pull request by fleverest**
This PR will resolve the first steps for #19.

So far the following has been done:
* Added a fully-functional R6 interface to wrap the Rcpp object, then did some plumbing so that the S3 interface utilizes calls to the R6 methods. Now the user has an option to use either the R6 or the S3 interface. Aside from renaming functions and parameters, there shouldn't be any other breaking changes.
* Added references using Rdpack.
* Added a GPL-3 license to the repository.
* Wrote more tests, specifically surrounding the internal distributions I implemented in C++.
* Restructured the Rcpp interface in the source.
* Added a print method for displaying useful information about the state of the posterior distribution.

### CODE REVIEW

**Comment by fleverest** [comment]
@dvukcevic What do you think of these changes? The main changes are an additional R6 interface which has some nice features like chaining commands:
```R
dtree <- dirichlet_tree$new(
  candidates = LETTERS[1:5],
  a0 = 1.,
  min_depth = 1,
  max_depth = 3,
  vd = FALSE
)$update(
  ranked_ballots(
    list(c("A", "B", "C"), c("E", "D", "C"))
  )
)
```
And I added a `print(dtree)` method which outputs a brief quality-of-life summary:

```
Dirichlet-Tree:
  Candidates: E B D C A

  `a0`: 1

  `min_depth`: 1

  `max_depth`: 3

  `vd`: FALSE

  observations: 2
    first preference distribution:

    A     B     C     D     E EMPTY
    1     0     0     0     1     0
```

In particular, I'm wondering if you have any further ideas for this print method. Perhaps there's a nicer way to format the output? Maybe there's some more information which might be useful?

If interested, you can take a look at the full `dirichlet_tree` R6 documentation by loading the package with devtools and running `Rdpack::viewRd("man/dirichlet_tree.Rd")`.

**Comment by fleverest** [comment]
TODO: compress print statement parameters to a single line

**Comment by fleverest** [comment]
I just updated the `print` formatting, so that

```R
print(
  dirichlet_tree$new(
    candidates = LETTERS[1:5],
    a0 = 1.,
    min_depth = 1,
    max_depth = 3,
    vd = FALSE
  )$update(ranked_ballots(list(c("A", "B", "C"), c("E", "D", "C"))))
)
```
prints the following:
```
Dirichlet-Tree:
  Candidates: A B C D E

  `a0`: 1, `min_depth`: 1, `max_depth`: 3, `vd`: FALSE

  observations: 2
    First preferences:
    A     B     C     D     E EMPTY
    1     0     0     0     1     0
```

I also added a pre-commit configuration which enforces strict styling and requires `roxygen2` documentation generation before committing among other things.

File: R/ballots.R [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+
+  # Return the ranked_ballots object
+  return(structure(
+    x,
+    class = "ranked_ballots",
+    candidates = candidates
+  ))
+}
+
+#' @name write.ballots
```
Comment (dvukcevic) [inline R/ballots.R:93]: At this line, we should have `write_ballots` rather than `write.ballots`

**Comment by dvukcevic** [comment]
I like the new, compact `print` output.

Can we simplify it even further?
How about something like this:

```
Dirichlet-tree (a0=1, min_depth=1, max_depth=3, vd=FALSE)
  Candidates: A B C D E
  Observations: 2
    First preferences:
    A     B     C     D     E EMPTY
    1     0     0     0     1     0
```

**Comment by fleverest** [comment]
I believe this can merge now.

**Comment by dvukcevic** [comment]
Looking great so far.  I've re-run my functional tests and they look fine with the latest version here.

However, we still need to review and update the documentation before it's ready for release.

Here are a few things I noted:

- [x] Add installation instructions in the README.
- [x] Change any instance of first-person to third-person, e.g. in the README.
- [x] Consistent capitalisation: always write 'Dirichlet-tree'.  (The only exception would be where it is part of a title and you are using maximal capitalisation, but I would actually recommend switching all titles to minimal capitalisation.)
- [ ] Add a website using [pkgdown](https://pkgdown.r-lib.org/).
- [ ] Cite our upcoming paper, once we release it (should be within the coming week).

**Comment by dvukcevic** [comment]
Suggested wording for the 2nd paragraph in the README:

The Dirichlet distribution is a simple and effective choice if the number of ballot types is small, but becomes problematic when the number gets large.  As `n` grows, the prior concentration parameters (defined by `a0` in our implementation) would need to be on the order of `1/n!` to ensure the prior is not overly informative.  If `n` is large enough, this would be smaller than floating-point precision would allow.  Also, the fact that this varies by `n` makes it inconvenient.  A more practical alternative is given by the Dirichlet-tree distribution, which we implement in this package.

**Comment by fleverest** [comment]
@dvukcevic Thank you for the checklist! I'm going to merge this PR now. I'll work on the `pkgdown` site with a new branch and I'll add the new references as soon as they're up on arxiv.

**Pull request merged**